requirements.txt

```diff
@@ -0,0 +1,57 @@
+antlr4-python3-runtime==4.9.2
+attrs==20.3.0
+biolinkml==1.7.6
+bmt==0.5.0
+cachetools==4.2.1
+certifi==2020.12.5
+CFGraph==0.2.1
+chardet==4.0.0
+click==7.1.2
+deprecation==2.1.0
+fastapi==0.63.0
+frozendict==1.2
+graphviz==0.16
+h11==0.9.0
+hbreader==0.9.1
+httpcore==0.12.3
+httptools==0.1.2
+httpx==0.16.1
+idna==2.10
+isodate==0.6.0
+jsonasobj==1.3.1
+jsonschema==3.2.0
+lxml==4.6.3
+packaging==20.9
+parse==1.19.0
+pbr==5.6.0
+prefixcommons==0.1.9
+prologterms==0.0.6
+pydantic==1.8.1
+PyJSG==0.11.6
+PyLD==2.0.3
+pyparsing==2.4.7
+pyrsistent==0.17.3
+PyShEx==0.7.20
+PyShExC==0.8.3
+PyYAML==5.4.1
+rdflib==5.0.0
+rdflib-jsonld==0.5.0
+rdflib-pyld-compat==0.1.0
+reasoner-pydantic @ git+git://github.com/ranking-agent/reasoner-pydantic@79be5c1e32aacc1d502f7071d518a4994afba04e
+redis==3.5.3
+requests==2.25.1
+rfc3986==1.4.0
+ShExJSG==0.7.1
+six==1.15.0
+sniffio==1.2.0
+sparqlslurper==0.4.1
+SPARQLWrapper==1.8.5
+starlette==0.13.6
+stringcase==1.2.0
+testfixtures==6.17.1
+typing-extensions==3.7.4.3
+urllib3==1.26.4
+uvicorn==0.11.7
+uvloop==0.15.2
+watchdog==2.0.3
+websockets==8.1
```

Comment:
We don't use all of these yet, and I'm inclined to add them as needed.

Reply:
It seems like this exposes a problem with the way we are using a fully specified requirements file - it makes it hard to remove packages + their dependencies. There is [a way](https://www.activestate.com/resources/quick-reads/how-to-uninstall-python-packages/) to uninstall a package and it's dependencies but that may break other installed packages.

Let me know if you have ideas. I think that having a `requirements.txt` and `requirements-lock.txt` file makes the most sense.

Reply:
Virtual environments are pretty easy to set up and tear down. If we need to remove a package, can't we just build a new venv and install only the packages we need? As they are installed, pip will enforce consistency.

Reply:
Maybe I'm misunderstanding something about this workflow. I'm comparing this to something like npm. Let's say I need to install fastapi. I write:
```
npm install fastapi
```
then later to uninstall I write:
```
npm uninstall fastapi
```

which removes fastapi and dependencies from the package.json and package-lock.json. I'm trying to figure out what a similar workflow is to manage this with a single requirements.txt file. I create a virtual environment then run:
```
pip install -r requirements.txt
pip install fastapi
pip freeze > requirements.txt
```

This adds `fastapi` + dependencies to requirements.txt (pydantic, starlette, typing-extensions). Later if I want to uninstall I can rebuild the venv and then:
```
pip install -r requirements.txt
pip uninstall fastapi
pip freeze > requirements.txt
```

The problem is that this leaves pydantic, starlette, and typing-extensions in the requirements.txt file. I can't figure out a way to cleanly remove those dependencies.

Reply:
Ahhhh, I think I finally understand the issue. We have no way to re-build the venv without a specific package because we don't know the dependence tree. So we need somewhere a list of our direct dependencies only. That's the `requirements.txt` file, so we can just remove the unwanted dependency from `requirements.txt` then
```bash
pip install -r requirements.txt
pip freeze > requirements-lock.txt
```
And `requirements-lock.txt` is what we'll use for testing and deployment. Thanks for your patience, @alongreyber.